text/0028-bill-of-materials.md

```diff
@@ -0,0 +1,166 @@
+# Implement a Bill of Materials Across Paketo
+
+## Summary
+
+A bill of materials (BOM) is a set of metadata that provides an account of the
+dependencies and tools that are both present in an application image, as well
+those used in the construction of that image. The availability of this data is
+a key value proposition of Cloud Native Buildpacks.
+
+As an implementation of Cloud Native Buildpacks, the Paketo project should
+support adding bill of materials metadata across the different components that
+go into an app image. This RFC serves as an intent to implement the BOM and
+defines the minimal set of key/values to be included.
+
+## Motivation
+
+A BOM can include an array of different metadata about application dependencies
+and buildpack dependencies such as SHAs, URIs, versions, license information,
+vulnerability identifiers, etc. This information is valuable for users who want
+an accurate inventory of everything in their final application image, or that
+may have been used to build it. Having access to that information makes it
+easier to audit for vulnerabilities and license information.
+
+The Paketo Java Buildpacks as well as a handful of other Paketo Buildpacks
+already make the BOM available. Another motivation for doing this work is to
+achieve a consistent user experience across the Paketo project.
+
+## Detailed Explanation
+
+The following Paketo components should have BOM metadata attached to them:
+
+### Overall Schema
+The bill of materials for each of type of entry should have some subset of
+fields that conform to the overall schema below.
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
+  version = "<dependency version>"
+  sha256 = "<hash of dependency artifact from uri>"
+  uri = "<uri to dependency>"
+  arch = "<compatible architecture>"
+  summary = "<package summary>"
+  cpe = "<version-specific common platfrom enumeration>"
+  purl = "<package URL>"
+  licenses = [<licenses that the dependency has>]
+
+[bom.metadata.source]
+  name = "<dependency source name>"
+  version = "<dependency source version>",
+  uri = "<uri to the dependency source>"
+  sha256 = "<hash of the dependency source artifact from source-uri>"
+  upstream-version = "<dependency source upstream version>"
+```
+
+### Stacks
+Stacks (such as those found in the [Paketo Stacks
+repository](https://github.com/paketo-buildpacks/stacks)) should have BOM
+metadata that includes in-depth information on all of the OS level packages
+installed as part of the stack.
+
+The minimal set of keys to include in these type of BOM entries are:
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
+  version = "<dependency version>"
+
+[bom.metadata.source]
+  name = "<dependency source name>"
+  version = "<dependency source version>",
+  upstream-version = "<dependency source upstream version>"
+```
+The optional set of keys may include:
```

Comment:
Is there any reason to call out optional keys? I was assuming anything in the schema that wasn't required was optional.

Reply:
Good call. Addressed in https://github.com/paketo-buildpacks/rfcs/pull/57/commits/7575872a10d60d0017ce9d052ca34a534bcde553

Reply:
Sorry for nitpicking on this but I am still confused after the above linked changes. In my mind there are only two things we need to specify. The complete schema (any field in which could optionally be filled out, for any type of entry) and any required fields for specific entry types. I don't understand why we re repeating subsets of the schema if to indicate required fields.

Are these the suggested fields (in which case maybe we could teak the wording to make that clear)? Or does this imply that other fields in the schema are disallowed for these dependency types (I don't think we should do this)?

Reply:
Following this comment/our discussion around this at the 7-27 Working Group meeting, we have gone ahead and streamlined the RFC to solely include the top-level schema to show possible fields, and removed the dependency-type schemas. They are "suggested fields", the only required fields are "name" and "version".